packages/flutter/lib/src/painting/_image_resolution_web.dart

```diff
@@ -0,0 +1,179 @@
+        // ourselves.
+        result = SynchronousFuture<AssetBundleImageKey>(key);
+      }
+    }).catchError((dynamic error, StackTrace stack) {
+      // We had an error. (This guarantees we weren't called synchronously.)
+      // Forward the error to the caller.
+      assert(completer != null);
+      assert(result == null);
+      completer.completeError(error, stack);
+    });
+    if (result != null) {
+      // The code above ran synchronously, and came up with an answer.
+      // Return the SynchronousFuture that we created above.
+      return result;
+    }
+    // The code above hasn't yet run its "then" handler yet. Let's prepare a
+    // completer for it to use when it does run.
+    completer = Completer<AssetBundleImageKey>();
+    return completer.future;
+  }
+
+  static Future<Map<String, List<String>>> _manifestParser(String jsonData) {
+    if (jsonData == null) {
+      return null;
+    }
+    // TODO(ianh): JSON decoding really shouldn't be on the main thread.
+    final Map<String, dynamic> parsedJson = json.decode(jsonData);
+    final Iterable<String> keys = parsedJson.keys;
+    final Map<String, List<String>> parsedManifest =
+        Map<String, List<String>>.fromIterables(
+            keys,
+            keys.map<List<String>>(
+                (String key) => List<String>.from(parsedJson[key])));
+    // TODO(ianh): convert that data structure to the right types.
+    return SynchronousFuture<Map<String, List<String>>>(parsedManifest);
+  }
+
+  String _chooseVariant(
+      String main, ImageConfiguration config, List<String> candidates) {
+    if (config.devicePixelRatio == null ||
+        candidates == null ||
+        candidates.isEmpty) {
+          return main;
+    }
+    // TODO(ianh): Consider moving this parsing logic into _manifestParser.
+    final SplayTreeMap<double, String> mapping = SplayTreeMap<double, String>();
+    for (String candidate in candidates)
+      mapping[_parseScale(candidate)] = candidate;
+    // TODO(ianh): implement support for config.locale, config.textDirection,
+    // config.size, config.platform (then document this over in the Image.asset
+    // docs)
+    return _findNearest(mapping, config.devicePixelRatio);
+  }
+
+  // Return the value for the key in a [SplayTreeMap] nearest the provided key.
+  String _findNearest(SplayTreeMap<double, String> candidates, double value) {
+    if (candidates.containsKey(value)) {
+      return candidates[value];
+    }
+    final double lower = candidates.lastKeyBefore(value);
+    final double upper = candidates.firstKeyAfter(value);
+    if (lower == null) {
+      return candidates[upper];
+    }
+    if (upper == null) {
+      return candidates[lower];
+    }
+    if (value > (lower + upper) / 2)
+      return candidates[upper];
+    else
+      return candidates[lower];
+  }
+
+  static final RegExp _extractRatioRegExp = RegExp(r'/?(\d+(\.\d*)?)x$');
+
+  double _parseScale(String key) {
```

Comment:
It seems like this might be the only difference? Is this change back-portable to flutter proper?

Reply:
I believe so. This code might not work on windows since it uses '/' as a path separator. Unless Flutter's asset keys always use '/' as the path separator even on Windows.

Reply:
I believe they are file uris and not paths, at least I hope so ....